service/awsconfig/v8/resource/lifecycle/current.go

```diff
@@ -95,34 +101,212 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(customObject)),
 		}
 
-		_, err := r.aws.AutoScaling.CompleteLifecycleAction(i)
+		nodeConfigs, err := r.g8sClient.CoreV1alpha1().NodeConfigs(n).List(o)
 		if err != nil {
 			return nil, microerror.Mask(err)
 		}
 
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("completed lifecycle hook action for guest cluster instance '%s'", *instance.InstanceId))
+		if len(nodeConfigs.Items) > 0 {
+			for _, nodeConfig := range nodeConfigs.Items {
+				r.logger.LogCtx(ctx, "level", "debug", "message", "inspecting node config for the guest cluster")
+
+				if hasFinalStatus(nodeConfig.Status.Conditions) {
+					r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has final state")
+				} else {
+					r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has no final state")
+					continue
+				}
+
+				instanceID, err := instanceIDFromAnnotations(nodeConfig.GetAnnotations())
+				if err != nil {
+					return nil, microerror.Mask(err)
+				}
+
+				err = r.completeLifecycleHook(ctx, instanceID, workerASGName)
+				if err != nil {
+					return nil, microerror.Mask(err)
+				}
+
+				err = r.deleteNodeConfig(ctx, nodeConfig.GetName())
+				if err != nil {
+					return nil, microerror.Mask(err)
+				}
+
+				r.logger.LogCtx(ctx, "level", "debug", "message", "inspected node config for the guest cluster")
+			}
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "no node configs to inspect for the guest cluster")
+		}
 	}
 
 	return nil, nil
 }
+
+func (r *Resource) completeLifecycleHook(ctx context.Context, instanceID, workerASGName string) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("completing lifecycle hook action for guest cluster instance '%s'", instanceID))
+
+	i := &autoscaling.CompleteLifecycleActionInput{
+		AutoScalingGroupName:  aws.String(workerASGName),
+		InstanceId:            aws.String(instanceID),
+		LifecycleActionResult: aws.String("CONTINUE"),
+		LifecycleHookName:     aws.String(key.NodeDrainerLifecycleHookName),
+	}
+
+	_, err := r.aws.AutoScaling.CompleteLifecycleAction(i)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("completed lifecycle hook action for guest cluster instance '%s'", instanceID))
+
+	return nil
+}
+
+func (r *Resource) createNodeConfig(ctx context.Context, customObject providerv1alpha1.AWSConfig, instanceID, privateDNS string) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "creating node config for guest cluster node")
+
+	n := v1.NamespaceDefault
+	c := &v1alpha1.NodeConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				key.InstanceIDAnnotation: instanceID,
+			},
+			Labels: map[string]string{
+				key.ClusterIDLabel: key.ClusterID(customObject),
+			},
+			Name: privateDNS,
```

Comment:
The unique identifier for our node configs is the actual node name which on AWS is defined by the private DNS.

Reply:
Yeah, this name was strange. Took me a while to realize it's in fact `nodeID` - I would call it like that, what value it actually holds it's less important.

Reply:
I would like to keep this but add comments instead. Does that work for you?

Reply:
It works for me already.